### PR TITLE
wtxmgr: resolve duplicate entries for outputs within bucketUnspent

### PR DESCRIPTION
**Pull request by wpaulino**
In this PR, we resolve a lingering bug within the wallet where it's possible that an output is added as unconfirmed credit after the fact that it has already confirmed. This would lead to duplicate entries for the same output within the wallet causing double spend transactions to be crafted.

The root cause for the unconfirmed credit being added again is most likely triggered by receiving an unconfirmed `chain.RelevantTx` notification from the backend after the transaction has already confirmed. Ideally this should not happen, but that fix is a lot more involved.

Fixes https://github.com/lightningnetwork/lnd/issues/1494.

### CODE REVIEW

File: wtxmgr/tx_test.go [hunk 1]
```diff
@@ -1724,3 +1725,138 @@ func TestInsertConfirmedDoubleSpendTx(t *testing.T) {
...
+
+// TestInsertDuplicateCreditAfterConfirm aims to test the case where a duplicate
+// unconfirmed credit is added to the store after the intial credit has already
+// confirmed. This can lead to outputs being duplicated in the store, which can
+// lead to creating double spends when querying the wallet's UTXO set.
+func TestInsertDuplicateCreditAfterConfirm(t *testing.T) {
+	t.Parallel()
+
+	store, db, teardown, err := testStore()
+	defer teardown()
```
Comment (cfromknecht) [inline wtxmgr/tx_test.go]: this should come after error check

Comment (wpaulino) [reply wtxmgr/tx_test.go]: Fixed.

File: wtxmgr/tx.go [hunk 2]
```diff
@@ -405,10 +405,15 @@ func (s *Store) AddCredit(ns walletdb.ReadWriteBucket, rec *TxRecord, block *Blo
 // duplicate (false).
 func (s *Store) addCredit(ns walletdb.ReadWriteBucket, rec *TxRecord, block *BlockMeta, index uint32, change bool) (bool, error) {
 	if block == nil {
+		// Skip from adding the credit if it's already been seen either
```
[the remark this answers is not in the document]
Comment (wpaulino) [reply wtxmgr/tx.go]: Fixed.

File: wtxmgr/tx_test.go [hunk 3]
```diff
@@ -1728,3 +1728,137 @@ func TestInsertConfirmedDoubleSpendTx(t *testing.T) {
...
+	spendTx := spendOutput(&cbRec.Hash, 0, 5e7, 4e7)
+	spendTxRec, err := NewTxRecordFromMsgTx(spendTx, b101.Time)
+	if err != nil {
+		t.Fatal(err)
+	}
+	commitDBTx(t, store, db, func(ns walletdb.ReadWriteBucket) {
+		if err := store.InsertTx(ns, spendTxRec, nil); err != nil {
+			t.Fatal(err)
+		}
+		err := store.AddCredit(ns, spendTxRec, nil, 1, true)
```
Comment (halseth) [inline wtxmgr/tx_test.go]: question: is `InsertTx` always followed by an `AddCredit`? If yes maybe they could be combined?

File: wtxmgr/query_test.go [hunk 4]
```diff
@@ -525,31 +534,23 @@ func TestStoreQueries(t *testing.T) {
...
-	newState.blocks = [][]TxDetails{
-		newState.blocks[0],
-		{newState.blocks[1][0]},
-		{newState.blocks[1][1]},
-	}
-	newState.blocks[1][0].Block = b101
-	newState.txDetails[recA.Hash][1].Block = b101
-	lastState = newState
-	tests = append(tests, queryTest{
-		desc:    "mine duplicate tx A",
```
Comment (halseth) [inline wtxmgr/query_test.go]: what happened to this test?

**Comment by halseth** [comment]
This is an awesome find! Should hopefully get rid of the double spending issues we have been seeing! 👍

File: wtxmgr/tx_test.go [hunk 5]
```diff
@@ -64,17 +64,18 @@ func testStore() (*Store, walletdb.DB, func(), error) {
...
 		return nil, nil, func() {}, err
 	}
+
 	db, err := walletdb.Create("bdb", filepath.Join(tmpDir, "db"))
 	if err != nil {
-		teardown := func() {
-			os.RemoveAll(tmpDir)
-		}
-		return nil, nil, teardown, err
+		os.RemoveAll(tmpDir)
```
Comment (cfromknecht) [inline wtxmgr/tx_test.go]: 👍

File: wtxmgr/tx.go [hunk 6]
```diff
@@ -405,10 +405,16 @@ func (s *Store) AddCredit(ns walletdb.ReadWriteBucket, rec *TxRecord, block *Blo
...
 func (s *Store) addCredit(ns walletdb.ReadWriteBucket, rec *TxRecord, block *BlockMeta, index uint32, change bool) (bool, error) {
 	if block == nil {
+		// If the outpoint that we should mark as credit already exists
+		// within the store, either as unconfirmed or confirmed, then we
+		// have nothing left to do and can exit.
 		k := canonicalOutPoint(&rec.Hash, index)
 		if existsRawUnminedCredit(ns, k) != nil {
 			return false, nil
 		}
+		if existsRawUnspent(ns, k) != nil {
```
Comment (cfromknecht) [inline wtxmgr/tx.go]: great catch

File: wtxmgr/query_test.go [hunk 7]
```diff
@@ -512,38 +512,6 @@ func TestStoreQueries(t *testing.T) {
...
-		{
-			TxRecord: *stripSerializedTx(recA),
-			Block:    makeBlockMeta(-1),
-		},
-		newState.blocks[1][0],
-	}
-	newState.txDetails[recA.Hash][0].Block = b100
-	newState.txDetails[recA.Hash] = append(newState.txDetails[recA.Hash], newState.blocks[1][0])
-	lastState = newState
-	tests = append(tests, queryTest{
```
Comment (cfromknecht) [inline wtxmgr/query_test.go]: why was this removed?

File: wtxmgr/query_test.go (hunk 4 above)
Comment (wpaulino) [reply wtxmgr/query_test.go]: Previously, the store would allow storing the same hash in both unconfirmed and confirmed buckets at the same time. We no longer allow this so the test had to be removed, otherwise it'd cause a failure.

File: wtxmgr/query_test.go (hunk 7 above)
Comment (wpaulino) [reply wtxmgr/query_test.go]: See https://github.com/btcsuite/btcwallet/pull/534#discussion_r214225146.

File: wtxmgr/tx_test.go (hunk 3 above)
Comment (wpaulino) [reply wtxmgr/tx_test.go]: Do you mean just combine them for the sake of the test? Outside of tests, it looks like it'd be a pretty big, possibly irrelevant change. FWIW there's a TODO for it https://github.com/btcsuite/btcwallet/blob/master/wtxmgr/tx.go#L387.
